hashjoin.go

```diff
@@ -0,0 +1,393 @@
+		valCol := flow.b[colIdx].(intColumn)[:batchRowLen]
+
+		for i := 0; i < nResults; i++ {
+			outCol[i] = valCol[resultsIdx[i]]
+		}
+	}
+
+	outFlow.n = nResults
+}
+
+// hashJoinGroupBuilder is an alternative builder that pre-loads all build
+// relation batches before building the hash table.
+type hashJoinGroupBuilder struct {
+	hashTable *hashTableInt
+
+	totalSize int
+
+	// bucket stores the hashed value of every key in the build relation, and
+	// thereby contributes additional memory in comparison to the
+	// hashJoinBuilder.
+	bucket []int
+}
+
+func (hashTable *hashTableInt) growGroup(amount int) {
+	hashTable.keys = append(hashTable.keys, make(intColumn, amount)...)
+
+	for i := 0; i < hashTable.nOutCols; i++ {
+		hashTable.values[i] = append(hashTable.values[i], make(intColumn, amount)...)
+	}
+
+	hashTable.allocated += amount
+}
+
+func (hashTable *hashTableInt) growNext(amount int) {
+	hashTable.next = append(hashTable.next, make([]int, amount)...)
+}
+
+func (hashTable *hashTableInt) insertKey(hashKey int, id int) {
+	hashTable.count++
+	hashTable.next[id] = hashTable.first[hashKey]
+	hashTable.first[hashKey] = id
+}
+
+func makeHashJoinGroupBuilder(hashTable *hashTableInt) *hashJoinGroupBuilder {
+	return &hashJoinGroupBuilder{
+		hashTable: hashTable,
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertBatch(flow dataFlow, eqColIdx int, outCols []int, batchSize int) {
+	builder.hashTable.growGroup(batchSize)
+
+	eqCol := flow.b[eqColIdx].(intColumn)[:batchRowLen]
+
+	for i := 0; i < batchSize; i++ {
+		builder.hashTable.keys[i + builder.totalSize + 1] = eqCol[i]
+	}
+
+	for idx, colIdx := range outCols {
+		outCol := flow.b[colIdx].(intColumn)[:batchRowLen]
+		builder.spread(outCol, idx, batchSize)
+	}
+
+	builder.totalSize += batchSize
+}
+
+func (builder *hashJoinGroupBuilder) spread(outCol intColumn, valIdx int, batchSize int) {
+	valCol := builder.hashTable.values[valIdx]
+
+	for i:= 0; i < batchSize; i++ {
+		valCol[i + builder.totalSize + 1] = outCol[i]
+	}
+}
+
+func (builder *hashJoinGroupBuilder) computeBucket() {
+	builder.bucket = make([]int, builder.totalSize)
+
+	for i := 0; i < builder.totalSize; i++ {
+		builder.bucket[i] = builder.hashTable.hashInt(builder.hashTable.keys[i + 1])
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertKeys() {
+	builder.hashTable.growNext(builder.totalSize)
+
+	for i := 0; i < builder.totalSize; i++ {
+		builder.hashTable.insertKey(builder.bucket[i], i + 1)
+	}
+}
+
+func (builder *hashJoinGroupBuilder) start(source ExecOp, eqColIdx int, outCols []int) {
+	for{
+		flow := source.Next()
+		batchSize := flow.n
+
+		if batchSize == 0 {
+			break
+		}
+
+		builder.insertBatch(flow, eqColIdx, outCols, batchSize)
+	}
+
+	builder.computeBucket()
+	builder.insertKeys()
+}
+
+// hashJoinerIntInt performs a hashJoin joining on two integer columns where the
+// left table represents the build relation. It does not work with N - N joins.
+type hashJoinerIntInt struct {
```

Comment:
nit: I would stick this in the top of the file! since it's the first thing somebody will want to read. The rest of the stuff might even belong in a separate file.